5.2 multicore markdelay

stedolan opened the pull request:
This patch extends the existing single-domain-only markdelay patch to work with multiple domains, and (in separate commits) re-enables all the parallel tests and adds a new parallel test.

The idea is to add a new phase, `Phase_sweep_main`, prior to `Phase_sweep_and_mark_main`, in which only sweeping and not marking can be done. Since the roots have not yet been marked, it is safe to give new allocations the `UNMARKED` status during this phase, allowing them to be collected a cycle earlier than otherwise. Also, caml_modify need not do darkening during this phase. This was implemented for single-domain programs in #2348, and this PR extends it to multi-domain programs.

Since domains freely share new allocations, all must agree on phase transitions. So, when any domain finishes sweeping, it requests a transition to Phase_sweep_and_mark_main. Instead of being acted upon directly, this sets a flag to trigger the phase transition at the next minor GC. (The phase transition involves marking roots, which must be done with the minor heap empty, and the simplest way to achieve this is to wait until the next point when the minor heap naturally empties)

mshinwell commented:
(This was automatically closed by Github when `5.2-merge-staging` was merged.  I've re-opened and changed the base to `main`.)

NickBarnes commented on ocaml/runtime/major_gc.c:
```diff
@@ -1368,35 +1373,52 @@ static intnat ephe_sweep (caml_domain_state* domain_state, intnat budget)
-  caml_gc_phase = Phase_sweep_and_mark_main;
+  caml_domain_state* domain = Caml_state;
+
+  /* Ephemerons */
```
Would be clearer if all this ephemerons stuff was in a separate function. AIUI it's required before we start marking, so maybe `prepare_ephemerons_for_marking()` or something.

NickBarnes commented on ocaml/runtime/major_gc.c:
```diff
@@ -1368,35 +1373,52 @@ static intnat ephe_sweep (caml_domain_state* domain_state, intnat budget)
-                                      &work_unstarted,
-                                      WORK_STARTED)){
-        caml_scan_global_roots(&caml_darken, domain);
+    if (atomic_load_relaxed(&domain_global_roots_started) == WORK_UNSTARTED &&
```
The domain which wins this atomic scans the global roots. Needs a comment.

stedolan replied:
Done

NickBarnes commented on ocaml/runtime/major_gc.c:
```diff
@@ -153,7 +153,8 @@ static atomic_uintnat work_counter;
+  WORK_STARTED,
+  WORK_COMPLETE
 };
 static atomic_uintnat domain_global_roots_started;
```
Also I think this can be a static local in the appropriate root-scanning function.

stedolan replied:
Done

NickBarnes commented on ocaml/runtime/major_gc.c:
```diff
@@ -2042,9 +2056,14 @@ int caml_mark_stack_is_empty(void)
+static void empty_mark_stack (void)
 {
   while (!Caml_state->marking_done){
+    while (!caml_marking_started()) {
```
Add comment to justify this annoying-but-necessary while loop.

stedolan replied:
Done

NickBarnes commented on ocaml/runtime/minor_gc.c:
```diff
@@ -830,7 +830,13 @@ int caml_do_opportunistic_major_slice
-void caml_empty_minor_heap_setup(caml_domain_state* domain_unused) {
+void caml_empty_minor_heap_setup(caml_domain_state* domain_unused,
+                                 void* mark_requested) {
+  /* Check whether the mark phase has been requested */
```
Moar comment!! Moar!!!

stedolan replied:
Done!!!1

NickBarnes reviewed (commented):
This is all OK, but definitely needs some more comments in a few places (and possibly some better names for things), as commented.

stedolan commented on ocaml/runtime/major_gc.c:
```diff
@@ -319,7 +323,7 @@ static void ephe_todo_list_emptied (void)
   /* Since the todo list is empty, this domain does not need to participate in
    * further ephemeron cycles. */
-  atomic_fetch_add(&ephe_cycle_info.num_domains_todo, -1);
+  atomic_fetch_add_verify_ge0(&ephe_cycle_info.num_domains_todo, -1);
```
Yeah, agreed.

stedolan commented:
I rebased to current `main` and added another commit with review feedback. There are no functional changes in the new commit - it's entirely shuffling code around, renaming, and adding comments.

NickBarnes commented on ocaml/runtime/major_gc.c:
```diff
@@ -1399,36 +1409,35 @@ void caml_mark_roots_stw (int participant_count, caml_domain_state** barrier_par
-    if (atomic_load_relaxed(&domain_global_roots_started) == WORK_UNSTARTED &&
-        atomic_compare_exchange_strong(&domain_global_roots_started,
+    if (atomic_load_relaxed(&global_roots_scanned) == WORK_UNSTARTED &&
+        atomic_compare_exchange_strong(&global_roots_scanned,
```
I rely on your knowledge of the C memory model etc: are we guaranteed that _some_ domain succeeds at this CAS? Or is it possible for them all to fail, in which case we need something else here....

stedolan replied:
Yes, you're guaranteed that if the CAS fails then a value other than `WORK_UNSTARTED` must have been read from `global_roots_scanned`, which means some other domain started successfully.

(This is what the `_strong` suffix means: there's also a `_weak` version that's allowed to fail spuriously, and may be faster on some platforms)

mshinwell reviewed (approved):
On behalf of Nick Barnes
